styx-common/src/main/java/com/spotify/styx/storage/CheckedDatastoreReaderWriter.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * -\-\-
+ * Spotify Styx Scheduler Service
+ * --
+ * Copyright (C) 2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.storage;
+
+import static com.google.common.collect.Iterables.toArray;
+
+import com.google.cloud.datastore.DatastoreException;
+import com.google.cloud.datastore.DatastoreReaderWriter;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.EntityQuery;
+import com.google.cloud.datastore.FullEntity;
+import com.google.cloud.datastore.Key;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * A wrapper for {@link DatastoreReaderWriter} that translates unchecked {@link DatastoreException}s to checked
+ * {@link DatastoreIOException}s.
+ */
+class CheckedDatastoreReaderWriter {
+
+  private final DatastoreReaderWriter rw;
+
+  CheckedDatastoreReaderWriter(DatastoreReaderWriter rw) {
+    this.rw = Objects.requireNonNull(rw);
+  }
+
+  Entity put(FullEntity<?> entity) throws IOException {
+    return call(() -> rw.put(entity));
+  }
+
+  Entity add(FullEntity<?> entity) throws IOException {
+    return call(() -> rw.add(entity));
+  }
+
+  void update(Entity entity) throws IOException {
+    run(() -> rw.update(entity));
+  }
+
+  Entity get(Key key) throws IOException {
+    return call(() -> rw.get(key));
+  }
+
+  List<Entity> get(Iterable<Key> keys) throws IOException {
+    return call(() -> ImmutableList.copyOf(rw.get(toArray(keys, Key.class))));
```

Comment:
At least this `get` is lazy according to Datastore javadoc, and `hasNext` or `next` may throw `DatastoreException`. :/

```
/**
   * Returns an {@link Entity} for each given {@link Key} that exists in the Datastore. The order of
   * the result is unspecified. Results are loaded lazily, so it is possible to get a
   * {@code DatastoreException} from the returned {@code Iterator}'s
   * {@link Iterator#hasNext hasNext} or {@link Iterator#next next} methods. {@link ReadOption}s can
   * be specified if desired.
   *
   * <p>Example of getting multiple entity objects.
   * <pre> {@code
   * String firstKeyName = "my_first_key_name";
   * String secondKeyName = "my_second_key_name";
   * KeyFactory keyFactory = datastore.newKeyFactory().setKind("MyKind");
   * Key firstKey = keyFactory.newKey(firstKeyName);
   * Key secondKey = keyFactory.newKey(secondKeyName);
   * Iterator<Entity> entitiesIterator = datastore.get(Lists.newArrayList(firstKey, secondKey));
   * List<Entity> entities = Lists.newArrayList();
   * while (entitiesIterator.hasNext()) {
   *   Entity entity = entitiesIterator.next();
   *   // do something with the entity
   *   entities.add(entity);
   * }
   * }</pre>
   *
   * @throws DatastoreException upon failure
   * @see #get(Key)
   */
```

Reply:
Yeah, so call `List<Entity> get(Iterable<Key> keys)` if you expect the results to be small, and call `void get(Iterable<Key> keys, IOConsumer<Entity> f)` if you expect results to be bigger and/or want to avoid collecting all entities into a list.

Both methods handle `hasNext ` and `next` throwing `DatastoreException`.

Reply:
Maybe we're missing some tests here proving that `DatastoreException` is handled.

Reply:
Shall we reflect the above comment about when to use what with some javadoc?